File: app/src/main/java/maderski/chargingindicator/sharedprefs/CIPreferences.java
package maderski.chargingindicator.sharedprefs;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Created by Jason on 6/11/16.
 */
public class CIPreferences {

    private static SharedPreferences.Editor _editor;

    private static final String MY_PREFS_NAME = "CIPreferences";

    private static final String VIBRATE_KEY = "vibrate";
    private static final String DISCONNECT_VIBRATE_KEY = "disconnectVibrateKey";
    private static final String DIFF_VIBRATIONS_KEY = "differentVibrationsKey";
    private static final String PLAY_SOUND_KEY = "playSound";
    private static final String DISCONNECT_PLAY_SOUND_KEY = "disconnectPlaySound";
    private static final String BATTERY_CHARGED_PLAY_SOUND_KEY = "batteryChargedPlaySound";
    private static final String SHOW_TOAST_KEY = "showToast";
    private static final String CHOSEN_DISCONNECT_SOUND_KEY = "chosenDisconnectSound";
    private static final String CHOSEN_CONNECT_SOUND_KEY = "chosenConnectSound";
    private static final String CHOSEN_BATTERY_CHARGED_SOUND_KEY = "batteryChargedSound";
    private static final String QUIET_TIME = "quietTime";
    private static final String START_QUIET_TIME = "startQuietTIme";
    private static final String END_QUIET_TIME = "endQuietTime";
    private static final String SHOW_CHARGING_BUBBLE_KEY = "showChargingBubble";
    private static final String CHARGING_BUBBLE_X = "chargingBubbleX";
    private static final String CHARGING_BUBBLE_Y = "chargingBubbleY";


    //Writes to SharedPreferences, but still need to commit setting to save it
    private static SharedPreferences.Editor editor(Context context){

        if(_editor == null){
            _editor = context.getSharedPreferences(MY_PREFS_NAME, Context.MODE_PRIVATE).edit();
            _editor.apply();
        }

        return _editor;
    }

    //Reads SharedPreferences value
    private static SharedPreferences reader(Context context){

        return context.getSharedPreferences(MY_PREFS_NAME, Context.MODE_PRIVATE);
    }

    //Commits write to SharedPreferences
    private static void commit(Context context){
        editor(context).commit();
        _editor = null;
    }

    public static void setChargingBubbleX(Context context, float x) {
        editor(context).putFloat(CHARGING_BUBBLE_X, x);
        commit(context);
    }

    public static float getChargingBubbleX(Context context){
        return reader(context).getFloat(CHARGING_BUBBLE_X, 60f);
    }

    public static void setChargingBubbleY(Context context, float y) {
        editor(context).putFloat(CHARGING_BUBBLE_Y, y);
        commit(context);
    }

    public static float getChargingBubbleY(Context context){
        return reader(context).getFloat(CHARGING_BUBBLE_Y, 20f);
    }

    public static void setShowChargingBubble(Context context, boolean showChargingFab) {
        editor(context).putBoolean(SHOW_CHARGING_BUBBLE_KEY, showChargingFab);
        commit(context);
    }

    public static boolean getShowChargingBubble(Context context){
        return reader(context).getBoolean(SHOW_CHARGING_BUBBLE_KEY, true);
    }

    public static void setStartQuietTime(Context context, int time){
        editor(context).putInt(START_QUIET_TIME, time);
        commit(context);
    }

    public static int getStartQuietTime(Context context){
        return reader(context).getInt(START_QUIET_TIME, 2200);
    }

    public static void setEndQuietTime(Context context, int time){
        editor(context).putInt(END_QUIET_TIME, time);
        commit(context);
    }

    public static int getEndQuietTime(Context context){
        return reader(context).getInt(END_QUIET_TIME, 800);
    }


    public static void setQuietTime(Context context, boolean enabled){
        editor(context).putBoolean(QUIET_TIME, enabled);
        commit(context);
    }

    public static boolean getQuietTime(Context context){
        return reader(context).getBoolean(QUIET_TIME, false);
    }

    public static void setBatteryChargedPlaySound(Context context, boolean enabled){
        editor(context).putBoolean(BATTERY_CHARGED_PLAY_SOUND_KEY, enabled);
        commit(context);
    }

    public static boolean getBatteryChargedPlaySound(Context context){
        return reader(context).getBoolean(BATTERY_CHARGED_PLAY_SOUND_KEY, false);
    }

    public static void setChosenBatteryChargedSound(Context context, String chosenSound){
        editor(context).putString(CHOSEN_BATTERY_CHARGED_SOUND_KEY, chosenSound);
        commit(context);
    }

    public static String getChosenBatteryChargedSound(Context context){
        return reader(context).getString(CHOSEN_BATTERY_CHARGED_SOUND_KEY, "None");
    }

    public static void setChosenDisconnectSound(Context context, String chosenSound){
        editor(context).putString(CHOSEN_DISCONNECT_SOUND_KEY, chosenSound);
        commit(context);
    }

    public static String getChosenDisconnectSound(Context context){
        return reader(context).getString(CHOSEN_DISCONNECT_SOUND_KEY, "None");
    }

    public static void setChosenConnectSound(Context context, String chosenSound){
        editor(context).putString(CHOSEN_CONNECT_SOUND_KEY, chosenSound);
        commit(context);
    }

    public static String getChosenConnectSound(Context context){
        return reader(context).getString(CHOSEN_CONNECT_SOUND_KEY, "None");
    }

    public static void setDisconnectPlaySound(Context context, boolean enabled){
        editor(context).putBoolean(DISCONNECT_PLAY_SOUND_KEY, enabled);
        commit(context);
    }

    public static boolean getDisconnectPlaySound(Context context){
        return reader(context).getBoolean(DISCONNECT_PLAY_SOUND_KEY, false);
    }

    public static void setVibrateOnDisconnect(Context context, boolean enabled){
        editor(context).putBoolean(DISCONNECT_VIBRATE_KEY, enabled);
        commit(context);
    }

    public static boolean getVibrateOnDisconnect(Context context){
        return reader(context).getBoolean(DISCONNECT_VIBRATE_KEY, false);
    }

    public static void setDiffVibrations(Context context, boolean enabled){
        editor(context).putBoolean(DIFF_VIBRATIONS_KEY, enabled);
        commit(context);
    }

    public static boolean getDiffVibrations(Context context){
        return reader(context).getBoolean(DIFF_VIBRATIONS_KEY, true);
    }

    public static void setVibrateWhenPluggedIn(Context context, boolean enabled){
        editor(context).putBoolean(VIBRATE_KEY, enabled);
        commit(context);
    }

    public static boolean GetVibrateWhenPluggedIn(Context context){
        return reader(context).getBoolean(VIBRATE_KEY, false);
    }

    public static void setPlaySound(Context context, boolean enabled){
        editor(context).putBoolean(PLAY_SOUND_KEY, enabled);
        commit(context);
    }

    public static boolean GetPlaySound(Context context){
        return reader(context).getBoolean(PLAY_SOUND_KEY, false);
    }

    public static void setShowToast(Context context, boolean enabled){
        editor(context).putBoolean(SHOW_TOAST_KEY, enabled);
        commit(context);
    }

    public static boolean GetShowToast(Context context){
        return reader(context).getBoolean(SHOW_TOAST_KEY, true);
    }
}

File: app/src/main/java/maderski/chargingindicator/sharedprefs/CIPreferences.kt
package maderski.chargingindicator.sharedprefs

import android.content.Context
import android.content.SharedPreferences
import maderski.chargingindicator.sharedprefs.CISettings.Companion.BATTERY_CHARGED_PERCENT
import maderski.chargingindicator.sharedprefs.CISettings.Companion.BATTERY_CHARGED_PLAY_SOUND_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.CHARGING_BUBBLE_X
import maderski.chargingindicator.sharedprefs.CISettings.Companion.CHARGING_BUBBLE_Y
import maderski.chargingindicator.sharedprefs.CISettings.Companion.CHOSEN_BATTERY_CHARGED_SOUND_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.CHOSEN_CONNECT_SOUND_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.CHOSEN_DISCONNECT_SOUND_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.DIFF_VIBRATIONS_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.DISCONNECT_PLAY_SOUND_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.DISCONNECT_VIBRATE_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.END_QUIET_TIME
import maderski.chargingindicator.sharedprefs.CISettings.Companion.PLAY_SOUND_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.QUIET_TIME
import maderski.chargingindicator.sharedprefs.CISettings.Companion.SHOW_CHARGING_BUBBLE_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.SHOW_TOAST_KEY
import maderski.chargingindicator.sharedprefs.CISettings.Companion.START_QUIET_TIME
import maderski.chargingindicator.sharedprefs.CISettings.Companion.VIBRATE_KEY

/**
 * Created by Jason on 6/11/16.
 */
object CIPreferences : CISettings {
    private const val MY_PREFS_NAME = "CIPreferences"
    private const val NONE = "None"

    override fun setBatteryChargedPercent(context: Context, percent: Int) = editor(context).putInt(BATTERY_CHARGED_PERCENT, percent).apply()
    override fun getBatteryCharged(context: Context): Int = reader(context).getInt(BATTERY_CHARGED_PERCENT, 100)

    override fun setChargingBubbleX(context: Context, x: Float) = editor(context).putFloat(CHARGING_BUBBLE_X, x).apply()
    override fun getChargingBubbleX(context: Context): Float = reader(context).getFloat(CHARGING_BUBBLE_X, 60f)

    override fun setChargingBubbleY(context: Context, y: Float) = editor(context).putFloat(CHARGING_BUBBLE_Y, y).apply()
    override fun getChargingBubbleY(context: Context): Float = reader(context).getFloat(CHARGING_BUBBLE_Y, 20f)

    override fun setShowChargingBubble(context: Context, showChargingFab: Boolean) = editor(context).putBoolean(SHOW_CHARGING_BUBBLE_KEY, showChargingFab).apply()
    override fun getShowChargingBubble(context: Context): Boolean = reader(context).getBoolean(SHOW_CHARGING_BUBBLE_KEY, false)

    override fun setStartQuietTime(context: Context, time: Int) = editor(context).putInt(START_QUIET_TIME, time).apply()
    override fun getStartQuietTime(context: Context): Int = reader(context).getInt(START_QUIET_TIME, 2200)

    override fun setEndQuietTime(context: Context, time: Int) = editor(context).putInt(END_QUIET_TIME, time).apply()
    override fun getEndQuietTime(context: Context): Int = reader(context).getInt(END_QUIET_TIME, 800)

    override fun setQuietTime(context: Context, enabled: Boolean) = editor(context).putBoolean(QUIET_TIME, enabled).apply()
    override fun getQuietTime(context: Context): Boolean = reader(context).getBoolean(QUIET_TIME, false)

    override fun setBatteryChargedPlaySound(context: Context, enabled: Boolean) = editor(context).putBoolean(BATTERY_CHARGED_PLAY_SOUND_KEY, enabled).apply()
    override fun getBatteryChargedPlaySound(context: Context): Boolean = reader(context).getBoolean(BATTERY_CHARGED_PLAY_SOUND_KEY, false)

    override fun setChosenBatteryChargedSound(context: Context, chosenSound: String) = editor(context).putString(CHOSEN_BATTERY_CHARGED_SOUND_KEY, chosenSound).apply()
    override fun getChosenBatteryChargedSound(context: Context): String = reader(context).getString(CHOSEN_BATTERY_CHARGED_SOUND_KEY, NONE) ?: NONE

    override fun setChosenDisconnectSound(context: Context, chosenSound: String) = editor(context).putString(CHOSEN_DISCONNECT_SOUND_KEY, chosenSound).apply()
    override fun getChosenDisconnectSound(context: Context): String = reader(context).getString(CHOSEN_DISCONNECT_SOUND_KEY, NONE) ?: NONE

    override fun setChosenConnectSound(context: Context, chosenSound: String) = editor(context).putString(CHOSEN_CONNECT_SOUND_KEY, chosenSound).apply()
    override fun getChosenConnectSound(context: Context): String = reader(context).getString(CHOSEN_CONNECT_SOUND_KEY, NONE) ?: NONE

    override fun setDisconnectPlaySound(context: Context, enabled: Boolean) = editor(context).putBoolean(DISCONNECT_PLAY_SOUND_KEY, enabled).apply()
    override fun getDisconnectPlaySound(context: Context): Boolean = reader(context).getBoolean(DISCONNECT_PLAY_SOUND_KEY, false)

    override fun setVibrateOnDisconnect(context: Context, enabled: Boolean) = editor(context).putBoolean(DISCONNECT_VIBRATE_KEY, enabled).apply()
    override fun getVibrateOnDisconnect(context: Context): Boolean = reader(context).getBoolean(DISCONNECT_VIBRATE_KEY, false)

    override fun setDiffVibrations(context: Context, enabled: Boolean) = editor(context).putBoolean(DIFF_VIBRATIONS_KEY, enabled).apply()
    override fun getDiffVibrations(context: Context): Boolean = reader(context).getBoolean(DIFF_VIBRATIONS_KEY, true)

    override fun setVibrateWhenPluggedIn(context: Context, enabled: Boolean) = editor(context).putBoolean(VIBRATE_KEY, enabled).apply()
    override fun getVibrateWhenPluggedIn(context: Context): Boolean = reader(context).getBoolean(VIBRATE_KEY, false)

    override fun setPlaySound(context: Context, enabled: Boolean) = editor(context).putBoolean(PLAY_SOUND_KEY, enabled).apply()
    override fun getPlaySound(context: Context): Boolean = reader(context).getBoolean(PLAY_SOUND_KEY, false)

    override fun setShowToast(context: Context, enabled: Boolean) = editor(context).putBoolean(SHOW_TOAST_KEY, enabled).apply()
    override fun getShowToast(context: Context): Boolean = reader(context).getBoolean(SHOW_TOAST_KEY, true)

    //Writes to SharedPreferences, but still need to commit setting to save it
    private fun editor(context: Context): SharedPreferences.Editor = context.getSharedPreferences(MY_PREFS_NAME, Context.MODE_PRIVATE).edit()

    //Reads SharedPreferences value
    private fun reader(context: Context): SharedPreferences = context.getSharedPreferences(MY_PREFS_NAME, Context.MODE_PRIVATE)

}
